demos/burgers-goal_oriented.py

```diff
@@ -0,0 +1,425 @@
+#
+# Looking at the final adapted mesh, we can make a few observations. Firstly, the mesh
+# elements are indeed isotropic. Secondly, the solution moves to the right, becoming
+# more densely distributed near to the right-hand boundary. This can be seen by
+# comparing the second mesh against the first.
+#
+# Recall that the Burgers problem is quasi-1D, since the analytical solution varies in
+# the :math:`x`-direction, but is constant in the :math:`y`-direction, suggesting a
+# more optimal choice of  goal-based error estimation for this problem is one which
+# accounts for this anisotopy.
+
+
+# Goalie also provides drivers for *anisotropic* goal-oriented mesh adaptation. Here,
+# we consider the ``anisotropic_dwr_metric`` driver. (See documentation for details.) To
+# use it, we just need to define a different adaptor function. The same error indicator
+# is used as for the isotropic approach. In addition, the Hessian of the forward
+# solution is estimated as in the `previous demo <.burgers-hessian.py.html>`__
+# to give anisotropy to the metric.
+#
+# For this driver, normalisation is handled differently than for ``isotropic_metric``,
+# where the ``normalise`` method is called after construction. In this case, the metric
+# is already normalised within the call to ``anisotropic_dwr_metric``, so this is not
+# required. ::
+#
+def adaptor(mesh_seq, solutions=None, indicators=None):
+    metrics = []
+    complexities = []
+
+    solutions = mesh_seq.solutions
+    indicators = mesh_seq.indicators
+
+    # Ramp the target average metric complexity per timestep
+    base, target, iteration = 400, 1000, mesh_seq.fp_iteration
+    mp = {
+        "dm_plex_metric": {
+            "target_complexity": ramp_complexity(base, target, iteration),
+            "p": 1.0,
+        }
+    }
+
+    # Construct the metric on each subinterval
+    for i, mesh in enumerate(mesh_seq):
+        sols = solutions["u"]["forward"][i]
+        dt = mesh_seq.time_partition.timesteps[i]
+
+        # Define the Riemannian metric
+        P1_ten = TensorFunctionSpace(mesh, "CG", 1)
+
+        # At each timestep, recover metric of the solution
+        # vector. Then time integrate over the contributions
+        metrics_i = []
+
+        # Calculate metric at each timestep
+        for j, sol in enumerate(sols):
+            # get local indicator
+            indi = indicators["u"][i][j]
+            # local instance of Riemanian metric
+            metric_j = RiemannianMetric(P1_ten)
+
+            # At each timestep, recover Hessians of the two components of the solution
+            # vector. Then time integrate over the contributions
+            hessians = [RiemannianMetric(P1_ten) for _ in range(2)]
+            for k, hessian in enumerate(hessians):
+                hessian.set_parameters(mp)
+                hessian.compute_hessian(sol[k])
+                hessian.enforce_spd(restrict_sizes=True)
+
+            # reset parameters as a precaution
```

Comment:
As in the previous `adaptor` function, I'd remove this comment and a few others here which are obvious or unnecessary :)